Command line static remapping (ROS1 compatible syntax + nodename prefix)

sloretz opened the pull request:
PR for CI + visibility. Not ready for review, though feedback would be appreciated.

This is an implementation of command line remapping using a subset of the ROS2 remapping grammar. It includes almost everything in ROS 1 (includes node names, node namespaces, topics, and services, missing parameter name remapping) plus node name prefixes from the ROS2 grammar. `rclcpp` nodes can be remapped as is. `rclpy` will need a PR to make it pass arguments to `rcl`.

```
ros2 run demo_nodes_cpp talker chatter:=converse
ros2 run demo_nodes_cpp listener chatter:=converse
```

example rules that work in this PR
```
__node:=new_node_name
old_node_name:__node:=new_node_name
__ns:=/remapped/namespace
foo:=bar
/foo/bar:=/bar/foo
```

TODO before this can be merged
* ~~fix crash using `__ns:=/remapped/`~~ fixed
* ~~rclpy PR to pass arguments~~ done ros2/rclpy#179 
* Split this PR into smaller chunks to make it more likely to be reviewed quickly
* Self-review code + looking for untested parts
*

CI (skipping examples_rclcpp_minimal_remapping)
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4033)](http://ci.ros2.org/job/ci_linux/4033/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1124)](http://ci.ros2.org/job/ci_linux-aarch64/1124/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3362)](http://ci.ros2.org/job/ci_osx/3362/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4125)](http://ci.ros2.org/job/ci_windows/4125/)

connects to ros2/ros2#450

wjwwood commented on rcl/src/rcl/remap.c:
```diff
@@ -0,0 +1,249 @@
+{
+  for (int i = 0; i < num_rules; ++i) {
+    rcl_remap_t * rule = &(remap_rules[i]);
+    if (rule->type & type_bitmask) {
```
suggestion: you could do something like this:

```
if (!(rule->type & type_bitmask)) {
  continue;
}
```

Which would allow you to unindent the rest of the while loop one level.

sloretz replied:
Less indentation and removed some unnecessary checks in 8cbe7f80e55451ec07a307d56af15e952ae95d8a

wjwwood commented on rcl/src/rcl/remap.c:
```diff
@@ -0,0 +1,249 @@
+          rcl_ret_t ret = rcl_expand_topic_name(
+            rule->match, node_name, node_namespace, substitutions, allocator, &expanded_match);
+          if (ret != RCL_RET_OK) {
+            // Expansion failed with this rule, try another
```
It would be worth considering why this failed, for example if it failed to allocate memory you might want to just stop rather than continue trying. On the other hand, if the match is just not a valid topic it might be safe to ignore and continue.

wjwwood replied:
At the very least you might need to do a `rcl_reset_error()` in cases where `rcl_expand_topic_name()` set the error state.

sloretz replied:
Better error handling in 8cbe7f80e55451ec07a307d56af15e952ae95d8a

Function returns right away if the node name or namespace is invalid, and if BAD_ALLOC comes up. That error code + message is given back to the caller of the remapping function called.

wjwwood commented on rcl/src/rcl/remap_impl.h:
```diff
@@ -0,0 +1,78 @@
+rcl_remap_t
+rcl_remap_get_zero_initialized();
+
+/// Deallocate memory used in an rcl_remap_t structure
```
Same comment about Deallocate being slightly misleading.

sloretz replied:
"reclaim" in 6ac891fd7a79a70e9f87926aefccbc48b379c7a3

wjwwood commented on rcl/src/rcl/remap_impl.h:
```diff
@@ -0,0 +1,78 @@
+rcl_ret_t
+rcl_remap_fini(
+  rcl_remap_t * rule,
+  rcl_allocator_t allocator);
```
Same comment about storing the allocator used rather than passing the same one twice in two places.

sloretz replied:
Allocator stored in `rcl_remap_t` in 6ac891fd7a79a70e9f87926aefccbc48b379c7a3

sloretz commented:
Thanks for the review @wjwwood! In general feel free to commit to any PR I open, though in this case I'd like to make the changes so that I have a better chance of remembering them in the future.

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+{
+#endif
+
+struct rcl_arguments_impl_t;
```
`rcl_command_line_arguments_t` is a pretty long name. How about `rcl_cli_arguments_t`?

wjwwood replied:
`rcl_cli_arguments_t` sounds great to me, but after thinking about it, I'm not really concerned about the name. I think in context it will be relatively clear. So I'll leave it up to you to decide what to do about it.

sloretz replied:
I'll leave this as is because I think the name is clear enough, but I'm open to changing it in another PR.

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+rcl_get_unparsed_arguments(
+  rcl_arguments_t * args,
+  rcl_allocator_t allocator,
+  int ** output_unparsed_indices);
```
Looks like either is grammatically fine. [dictionary.com](http://www.dictionary.com/browse/index), [merriam-webster](https://www.merriam-webster.com/dictionary/index), and [oxforddictionaries.com](https://en.oxforddictionaries.com/definition/index) list both `indexes` and `indices` as a plural form of index. The only differences I see is m-w says the plural form is usually `indices`, and oxford says `indexes` can also be a verb.

wjwwood replied:
Again, just a nitpick, nothing needs to be done here I think.

sloretz commented on rcl/include/rcl/remap.h:
```diff
@@ -0,0 +1,209 @@
+ * instructed to ignore global arguments.
+ *
+ * Remap rules are checked in the order they were given.
+ * Processing stops when a remap rule has been matched or there are no more rules.
```
> if not, should there be?

By multiple passes do you mean using every rule in order using the result of the previous rule rather than stopping when one matches?

An argument against is that remapping stopped at the first rule in ROS 1, so a set of rules that work in ROS 1 may not work in ROS 2. Another argument against is that it adds an extra rule to swap topic names: `foo:=temporary_name bar:=foo temporary_name:=bar`.

An argument for is that dynamic remapping could take advantage of it if publishers/clients/etc store the original name. Dynamic remapping would add a new rule to the end of the list and then have all rules re-applied. Names that come up after a dynamic rule is added could be remapped as if the dynamic rule was a static one all along.

sloretz replied:
I added the corner cases you mentioned and expanded the documentation in c04c2ebc7d21ec0a1d7fc99ff7d51320bcf22f07

wjwwood replied:
> By multiple passes do you mean using every rule in order using the result of the previous rule rather than stopping when one matches?

Yeah, that's what I meant, but I was thinking of a different way to describe it. Basically, you could continue applying rules, in order, to a single input until it matches none, but then you need to detect cycles...

I don't think we should support it, I just think it would be nice to point out that we don't do that explicitly. It's one of those things that someone could easily assume is happening without realizing that it doesn't because it's complicated and we didn't implement it.

I think the documentation changes you made are sufficient, thanks!

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+ *
+ * \param[in] argc the number of arguments in argv.
+ * \param[in] argv the values of the arguments.
+ * \param[in] allocator a valid allocator
```
Made punctuation consistent in 7963397d20a750a5d31265bae05d18eb6f0cdd3d

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+rcl_get_num_unparsed_arguments(
+  rcl_arguments_t * args);
+
+/// Return a list of indexes that weren't successfully parsed
```
7963397d20a750a5d31265bae05d18eb6f0cdd3d

sloretz commented on rcl/include/rcl/node.h:
```diff
@@ -65,6 +66,13 @@ typedef struct rcl_node_options_t
   rcl_allocator_t allocator;
+
+  /// If true then both global command line arguments and arguments below apply to this node
+  /// If false then only the local arguments apply
```
7963397d20a750a5d31265bae05d18eb6f0cdd3d

sloretz commented on rcl/include/rcl/remap.h:
```diff
@@ -0,0 +1,209 @@
+ * \param[in] allocator a valid allocator to use
+ * \param[out] output_name either an allocated string with the remapped name,
+ *             or `NULL` if no remap rules matched the name
+ * \return `RCL_RET_OK` if the topic name was remapped or no rules matched
```
7963397d20a750a5d31265bae05d18eb6f0cdd3d

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+ * \param[in] args an arguments structure that has been parsed.
+ * \param[in] allocator a valid allocator
+ * \param[out] output_unparsed_indices an allocated array of indices into the original argv array.
+ *  This array must be deallocated by the caller using the given allocator.
```
Consistent indentation in a98baf26df90e8dc02f9f24ed697fe77c14c4792

sloretz commented on rcl/include/rcl/remap.h:
```diff
@@ -0,0 +1,209 @@
+ * \param[in] node_namespace the namespace of a node to which name belongs
+ * \param[in] allocator a valid allocator to use
+ * \param[out] output_name either an allocated string with the remapped name,
+ *             or `NULL` if no remap rules matched the name
```
Indentation a98baf26df90e8dc02f9f24ed697fe77c14c4792

sloretz commented on rcl/include/rcl/remap.h:
```diff
@@ -0,0 +1,209 @@
+ * Remap rules are checked in the order they were given.
+ * Processing stops when a rule has been matched or there are no more rules.
+ *
+ *
```
Vertical whitespace minimized in a98baf26df90e8dc02f9f24ed697fe77c14c4792

sloretz commented on rcl/include/rcl/remap.h:
```diff
@@ -0,0 +1,209 @@
+ * Namespace remap rules are checked in the order they were given.
+ * Processing stops when a rule has been matched or there are no more rules.
+ *
+ *
```
a98baf26df90e8dc02f9f24ed697fe77c14c4792

sloretz commented on rcl/src/rcl/rcl.c:
```diff
@@ -33,6 +35,7 @@ static char ** __rcl_argv = NULL;
 static atomic_uint_least64_t __rcl_instance_id = ATOMIC_VAR_INIT(0);
 static uint64_t __rcl_next_unique_id = 0;
 
+
```
a98baf26df90e8dc02f9f24ed697fe77c14c4792

sloretz commented on rcl/include/rcl/remap.h:
```diff
@@ -0,0 +1,209 @@
+rcl_remap_topic_name(
+  const rcl_arguments_t * local_arguments,
+  bool use_global_arguments,
+  const char * name,
```
topic_name/service_name in a78266c665b6b49f2f4697a09eeae14683eded85

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+  rcl_allocator_t allocator,
+  int ** output_unparsed_indices);
+
+/// Deallocate a structure holding parsed command line arguments
```
`reclaim` in 6ac891fd7a79a70e9f87926aefccbc48b379c7a3

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+rcl_ret_t
+rcl_arguments_fini(
+  rcl_arguments_t * args,
+  rcl_allocator_t allocator);
```
Allocator is stored in `rcl_arguments_t` in 6ac891fd7a79a70e9f87926aefccbc48b379c7a3

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+{
+  RCL_CHECK_ALLOCATOR_WITH_MSG(&allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(args, RCL_RET_INVALID_ARGUMENT, allocator);
+  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Finalizing arguments");
```
Removed in 6ac891fd7a79a70e9f87926aefccbc48b379c7a3

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+  if (args->impl) {
+    if (args->impl->remap_rules) {
+      for (int i = 0; i < args->impl->num_remap_rules; ++i) {
+        rcl_remap_fini(&(args->impl->remap_rules[i]), allocator);
```
In 6ac891fd7a79a70e9f87926aefccbc48b379c7a3 it will log an error if it fails to finalize a remap rule, but continue trying to free the rest of the rules and the stuff in the struct. Also added `RCL_WARN_UNUSED` to `rcl_remap_fini`

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+
+    allocator.deallocate(args->impl, allocator.state);
+    args->impl = NULL;
+    RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Arguments finalized");
```
Removed in 6ac891fd7a79a70e9f87926aefccbc48b379c7a3

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+    RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Arguments finalized");
+    return RCL_RET_OK;
+  }
+  RCUTILS_LOG_WARN_NAMED(ROS_PACKAGE_NAME, "Arguments finalized_twice");
```
Done in 6ac891fd7a79a70e9f87926aefccbc48b379c7a3

sloretz commented on rcl/include/rcl/remap.h:
```diff
@@ -0,0 +1,209 @@
+{
+#endif
+
+/// Remaps a topic name based on given rules
```
Declarative brief descriptions in 6dbab0f8ffeb7f17827e025bcb6a3b655b71cbd7

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+// instance of global arguments
+rcl_arguments_t __rcl_arguments;
+
+/// \return true if c is in [a-zA-Z0-9_]
```
Fixed in 6dbab0f8ffeb7f17827e025bcb6a3b655b71cbd7

sloretz commented on rcl/src/rcl/arguments_impl.h:
```diff
@@ -0,0 +1,50 @@
+/// \sa rcl_init(int, char **, rcl_allocator_t)
+/// \internal
+RCL_LOCAL
+extern rcl_arguments_t __rcl_arguments;
```
Added `global` to the name, added `rcl_get_global_arguments()` , and made this a static variable in arguments.c in 015bff19a62597893e58635720c4529295030356

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+RCL_PUBLIC
+RCL_WARN_UNUSED
+int
+rcl_get_num_unparsed_arguments(
```
Sounds reasonable. I changed it to `rcl_arguments_get_count_unparsed()` to keep the length of the name down in 0ad0f6356c99ff50a1260bfe024c2e035078b30d

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_get_unparsed_arguments(
```
Changed to `rcl_arguments_get_unparsed()` in 0ad0f6356c99ff50a1260bfe024c2e035078b30d

sloretz commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,152 @@
+ * Lock-Free          | Yes
+ *
+ * \param[in] args an arguments structure that has been parsed.
+ * \return number of unparsed arguments, or -1 if args is invalid
```
`-1 if args is NULL or zero intialized` in 129aeb77ec6fa4633b5adfce7a2436a99f2021a1

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+#include "rcutils/logging_macros.h"
+#include "rcutils/strdup.h"
+#include "./arguments_impl.h"
+#include "./remap_impl.h"
```
Sorted in d08034b87d737c0d0f7c2a95645cf5c55748e415. I did the `./` includes above instead of separating them with a blank line because it's consistent with what was done in `rcl.c`

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+    goto fail;
+  }
+
+  // Attempt to parse arguments are remap rules
```
b52bb8bc13750ab4d631b34d57e7ea9cb8588ce1

sloretz commented on rcl/include/rcl/remap.h:
```diff
@@ -0,0 +1,209 @@
+ *
+ * \param[in] local_arguments command line arguments to be used before global arguments.
+ *            If NULL or zero-initialized then only global arguments are used.
+ * \param[in] use_global_arguments if false then global arguments aren't used at all
```
I like functions not using global state. Changed to pass global arguments in 0b465df212a58461d58ace7b90d55b62d473a63f

sloretz replied:
Done in 0b465df212a58461d58ace7b90d55b62d473a63f

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+      ret = RCL_RET_BAD_ALLOC;
+      goto fail;
+    }
+    args_impl->remap_rules = shrunk_rules;
```
It's a little shorter 21b98068672217e5a7bc14a600fbff8ef83e38a3

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+      ret = RCL_RET_BAD_ALLOC;
+      goto fail;
+    }
+    args_impl->unparsed_args = shrunk;
```
21b98068672217e5a7bc14a600fbff8ef83e38a3

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+fail:
+  fail_ret = ret;
+  if (NULL != args_impl) {
+    // assign to ret to suppress warning about not checking return of fini
```
Fair point. Logs an error in42ea8dffdde02672a9ccb76b0ff4c8d63df12a0b

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+  // A valid rule has two parts separated by :=
+  separator = strstr(arg, ":=");
+  if (NULL == separator) {
+    return RCL_RET_ERROR;
```
Improved error messages and return codes in 12e5f599606b9362dd4ce6d4ec4bdf9979f61606. I made the return code generic because this function non-public and only used in one place and I'm not sure if I'll need to change error messages when implementing the complete remapping grammar.

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+  len_match = separator - arg;
+  len_replacement = strlen(replacement_begin);
+  if (len_match <= 0 || len_replacement <= 0) {
+    return RCL_RET_ERROR;
```
12e5f599606b9362dd4ce6d4ec4bdf9979f61606

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+    // Replacement must be a valid topic name
+    char * copy_replacement = rcutils_strndup(replacement_begin, len_replacement, allocator);
+    if (NULL == copy_replacement) {
+      return RCL_RET_ERROR;
```
Fixed in 12e5f599606b9362dd4ce6d4ec4bdf9979f61606

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+      }
+      int validation_result;
+      size_t invalid_index;
+      rcl_ret_t ret = rcl_validate_topic_name(copy_replacement, &validation_result, &invalid_index);
```
Fixed in 12e5f599606b9362dd4ce6d4ec4bdf9979f61606

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+  } else if (RCL_NODENAME_REMAP == type) {
+    // Replacement may only be a token
+    for (size_t i = 0; i < len_replacement; ++i) {
+      if (!_rcl_valid_token_char(replacement_begin[i])) {
```
Nah it's perfect 12e5f599606b9362dd4ce6d4ec4bdf9979f61606

sloretz commented:
Not done addressing comments, but running CI to see if I've broken anything

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4064)](http://ci.ros2.org/job/ci_linux/4064/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1170)](http://ci.ros2.org/job/ci_linux-aarch64/1170/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3374)](http://ci.ros2.org/job/ci_osx/3374/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4139)](http://ci.ros2.org/job/ci_windows/4139/)

sloretz commented on rcl/include/rcl/remap.h:
```diff
@@ -0,0 +1,209 @@
+ * If no rules matched then global remap rules will be checked unless use_global_arguments is false.
+ *
+ * Remap rules are checked in the order they were given.
+ * Processing stops when a rule has been matched or there are no more rules.
```
Expanded documentation in c04c2ebc7d21ec0a1d7fc99ff7d51320bcf22f07

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+_rcl_valid_token_char(char c)
+{
+  // assumes ASCII
+  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_';
```
Function is gone in 1925221e827551261cb291a315d3fa517eb3de45

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+
+  // Make sure node name contains only valid characters
+  for (size_t i = 0; i < len_node_name; ++i) {
+    if (!_rcl_valid_token_char(arg[i])) {
```
Done in 1925221e827551261cb291a315d3fa517eb3de45

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+    size_t invalid_index;
+    rcl_ret_t ret = rcl_validate_topic_name(copy_replacement, &validation_result, &invalid_index);
+    if (ret != RCL_RET_OK || validation_result != RCL_TOPIC_NAME_VALID) {
+      ret = RCL_RET_ERROR;
```
Better error messages using `validation_result` in 4f5180f61271532718679934328d3a213b04f7ef. `rcl_parse_arguments` logs the error messages as `DEBUG`. I logged them this way to avoid unnecessary console output when passing in non-ros arguments.

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+    if (NULL == copy_match) {
+      return RCL_RET_ERROR;
+    }
+    ret = rcl_validate_topic_name(copy_match, &validation_result, &invalid_index);
```
Uses new method to avoid copy in 8cbe7f80e55451ec07a307d56af15e952ae95d8a

sloretz commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,377 @@
+    if (ret != RCL_RET_OK || validation_result != RCL_TOPIC_NAME_VALID) {
+      ret = RCL_RET_ERROR;
+    }
+    allocator.deallocate(copy_replacement, allocator.state);
```
All cases avoid the extra copy as of 8cbe7f80e55451ec07a307d56af15e952ae95d8a

sloretz commented:
@wjwwood I've addressed all your comments. The only two unchanged things is the name `rcl_arguments_t` and `indices`.

CI
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4086)](http://ci.ros2.org/job/ci_linux/4086/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1181)](http://ci.ros2.org/job/ci_linux-aarch64/1181/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3381)](http://ci.ros2.org/job/ci_osx/3381/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4146)](http://ci.ros2.org/job/ci_windows/4146/)

dirk-thomas commented on rcl/include/rcl/arguments.h:
```diff
@@ -0,0 +1,187 @@
+ * Lock-Free          | Yes
+ *
+ * \param[in] argc The number of arguments in argv.
+ * \param[in] argv Whe values of the arguments.
```
Spelling: `Whe`

sloretz replied:
Fixed in 74b042b53f01811c99588e3a2522c35a812ce6ce

dirk-thomas commented on rcl/src/rcl/arguments.c:
```diff
@@ -0,0 +1,403 @@
+  }
+
+  // Rule is valid, construct a structure for it
+  output_rule->allocator = allocator;
```
Should the function check if the pointer is no `NULL` before dereferencing it?

sloretz replied:
It's probably ok since the function is internal and only used in one place, but it's better have code be defensive so I added checks to this function in 01a13889c9669b9b8dfbd6777f0964601e44a5f8. I didn't find other unchecked arguments in this file. Mind commenting on the other places you saw?
